katsdpsigproc/accel.py

```diff
@@ -1436,3 +1437,116 @@ def _extract_slots(self, names, add_to_hidden):
+        shape = '×'.join(str(x) for x in slot.shape)
+        if not shape:
+            shape = 'scalar'
+        padded_shape = '×'.join(str(x) for x in slot.required_padded_shape())
+        dtype = str(np.dtype(slot.dtype))
+        return '<b>{name}</b><br/>{shape}<br/>{padded_shape}<br/>{dtype}'.format(
+            name=name, shape=shape, padded_shape=padded_shape, dtype=dtype)
+
+
+def _visualize_operation(g, operation, path, slot_map, no_detail_slots, edges):
+    """Recursive implementation of :func:`visualize_operation`.
+
+    Parameters
+    ----------
+    g : :class:`graphviz.Graph`
+        Graph or subgraph to be updated with the operation. If the operation has
+        children, it will be rendered as a subgraph, otherwise as a single slot
+        with ports for the slots.
+    operation : :class:`Operation`
+        Operation to add to the graph
+    path : tuple of str
+        Hierarchical name of `operation`. The last component is used for visual
+        display, and the rest for generating unique node labels.
+    slot_map : dict
+        Mapping from slots to node names (or port names). It is updated in place
+        with all the slots in this operation and its descendants.
+    no_detail_slots : set
+        Slots would should not show shape and dtype details. It is updated in
+        place with any descendant slots of this node.
+    edges : set
+        Edges to draw, as tuples of two :class:`IOSlotBase`s.
+    """
+
+    all_slots = {**operation.slots, **operation.hidden_slots}
+
+    if hasattr(operation, 'operations'):
+        # Graphviz treats subgraphs with the prefix 'cluster_' specially. We
+        # only want this for real subgraphs, not the root.
+        node_name = 'cluster_' + '!'.join(path) if len(path) > 1 else 'root'
+        for slot_name, slot in all_slots.items():
+            if isinstance(slot, CompoundIOSlot):
+                for child in slot.children:
+                    no_detail_slots.add(child)
+        with g.subgraph(name=node_name) as sub:
+            sub.attr('graph', label=path[-1])
+            for child_name, child in operation.operations.items():
+                _visualize_operation(sub, child, path + (child_name,),
+                                     slot_map, no_detail_slots, edges)
+            for slot_name, slot in list(all_slots.items()):
+                if slot not in slot_map:     # It's already represented in the child
+                    # ":" has special meaning to graphviz, so we replace it with "+".
+                    full_name = 'slot_{}_{}'.format('!'.join(path), slot_name.replace(':', '+'))
+                    slot_map[slot] = full_name
+                    label = '<' + _slot_label(slot, slot_name, slot in no_detail_slots) + '>'
+                    sub.node(full_name, label=label, shape='box')
+                else:
+                    del all_slots[slot_name]
+    else:
+        node_name = 'op_' + '!'.join(path)
+        label = io.StringIO()
```

Comment:
Why `label = io.StringIO(); label.write('...')` and not `label = ''; label += '...'`? (I guess that's one of  StringIO's strengths? Automatic line termination?)

Reply:
Building a string by repeated concatenation can potentially take quadratic time, so is generally an anti-pattern (although CPython has some optimisations for it). Using StringIO gives linear-time performance.

Reply:
Another option would be to collect the pieces and use `''.join` at the end. That might be a little more natural - what do you think?

Reply:
That makes sense! No preference - we can leave it as is.